bpo-39940: Micro-optimizations of PySequence_Tuple()

Marco-Sulla opened the pull request:
Added `new_n_tmp_1` and `new_n_tmp_2` for micro-optimization of `PySequence_Tuple()`

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-39940](https://bugs.python.org/issue39940) -->
https://bugs.python.org/issue39940
<!-- /issue-number -->

Commit:
micro-optimization of PySequence_Tuple()

Added `new_n_tmp_1` and `new_n_tmp_2` for micro-optimization of `PySequence_Tuple()`

Commit:
operators precedence

forgot that `+` has precedence over `>>`

Commit:
nano-optimization

Delayed definition of variables only when needed. 

(Well, it's not old-school style compliant, but I see this approach very frequently now...)

Commit:
minors

Commit:
fixed regression

`_PyTuple_Resize()` does not accept `NULL`

Commit:
minor

mpaolini commented on Objects/abstract.c:
```diff
@@ -1912,67 +1912,90 @@ PySequence_Tuple(PyObject *v)
+        PyObject *it = PyObject_GetIter(v);
+        
+        if (it == NULL) {
+            return NULL;
```
here you need to decref `result` otherwise the tuple reference will leak

Marco-Sulla replied:
Bella (Thank you)

Commit:
memory leak

fixed a memory leak caused by my code change, thanks to [mpaolini](https://github.com/python/cpython/pull/18703#discussion_r391300540)
